src/parse/core/routine.hpp

```diff
@@ -0,0 +1,156 @@
+#pragma once
+#include <cstring>
+#include <array>
+#include <utility>
+#include "state.hpp"
+
+namespace docgen {
+namespace parse {
+namespace core {
+
+// Tags indicate possible routines 
+enum class Routine {
+    READ = 0,
+    IGNORE_WS,
+    PROCESS,
+    NUM_ROUTINES
+};
+
+// Forward-declaration of a routine.
+// routine will parse the chunk of text starting at begin
+// until it reaches end, given current state, and update the json object.
+// If a state changes, routine will modify state.
+// When a state changes or the chunk has been fully read, the routine will finish.
+// It will always return the next routine to execute.
+// routine will perform differently depending on routine_tag.
+
+template <Routine routine_tag>
+struct routine
+{
+    template <class Cache>
+    static Routine run(Cache& cache, const char*& begin, const char* end);
+};
+
+namespace details {
+
+template <class Cache>
+inline Routine process_slash(Cache& cache, const char*& begin, const char* end)
+{
+    // symbol should always be equivalent to "/"
+    assert(cache.symbol.get()[0] == '/' && 
+           cache.symbol.get()[1] == '\0');
+
+    Routine next_routine = Routine::READ;
+
+    if (begin == end) {
+        return next_routine;
+    }
+
+    // single-line
+    if (*begin == '/') {
+        cache.state = State::SINGLE_LINE;
+        next_routine = Routine::IGNORE_WS;
+    }
+
+    // block
+    else if (*begin == '*') {
+        cache.state = State::BLOCK;
+        next_routine = Routine::IGNORE_WS;
+    }
+
+    cache.symbol.clear();
+    ++begin;
+
+    return next_routine;
+}
+
+} // namespace details
+
+// Read routine will read from begin until end change states
+// only to parse function/class declarations or single-line/block comments.
+// It is expected that symbol inside cache is either 
```

Comment:
What does "it is expected" mean? Is it undefined behavior otherwise? This hardly seems robust if so—undefined behavior is okay when you are just telling someone to pass something correctly, but when this requirement is layered under an abstracted data structure like this, it becomes messy/unclear.

Reply:
I put `assert` statement inside but haven't pushed.

Reply:
pushed